### PR TITLE
Improving frontier evaluation

### PR DESCRIPTION
**Pull request by ashwin2802**

### CODE REVIEW

**Comment by ashwin2802** [comment]
@gajena review ploxx

File: ariitk_planning_msgs/package.xml [hunk 1]
```diff
@@ -0,0 +1,18 @@
+<?xml version="1.0"?>
+<package format="2">
+  <name>ariitk_planning_msgs</name>
+  <version>0.0.0</version>
+  <description>The ariitk_planning_msgs package</description>
+
+  <maintainer email="ashwins@iitk.ac.in">Ashwin Shenai</maintainer>
+  <maintainer email="gajendarnagar02@gmail.com">Gajendra Nagar</maintainer>
+
+  <license>TODO</license>
```
Comment (gajena) [inline ariitk_planning_msgs/package.xml]: stil todo

File: frontier_explorer/cfg/params.yaml [hunk 2]
```diff
@@ -1,11 +1,13 @@
 # Frontier Exploration Parameters
```
Comment (gajena) [inline frontier_explorer/cfg/params.yaml:1]: unints pls for all

File: voxblox_global_planner/src/skeleton_global_planner.cpp [hunk 3]
```diff
@@ -193,6 +193,17 @@ double SkeletonGlobalPlanner::getMapDistance(
   return distance;
 }
 
+double SkeletonGlobalPlanner::getMapDistanceAndGradient(
```
Comment (gajena) [inline voxblox_global_planner/src/skeleton_global_planner.cpp]: why distance is return and gradient is by reference?

File: voxblox_global_planner/src/skeleton_global_planner.cpp [hunk 4]
```diff
@@ -193,6 +193,17 @@ double SkeletonGlobalPlanner::getMapDistance(
...
+    const Eigen::Vector3d& position, Eigen::Vector3d& gradient) const {
```
Comment (gajena) [inline voxblox_global_planner/src/skeleton_global_planner.cpp]: why method const does not give good vibes

File: frontier_explorer/src/frontier_evaluator_node.cpp [hunk 5]
```diff
@@ -18,8 +18,7 @@ int main(int argc, char** argv) {
```
Comment (gajena) [inline frontier_explorer/src/frontier_evaluator_node.cpp:21]: please make rate a param or run whenever map is recieved.

File: frontier_explorer/src/frontier_evaluator.cpp [hunk 6]
```diff
@@ -136,28 +211,31 @@ void FrontierEvaluator::createMarkerFromFrontiers(visualization_msgs::MarkerArra
...
+        cube_center.y = frontier.center.y();
+        cube_center.z = frontier.center.z();
+        center.points.push_back(cube_center);
+    }
     markers->markers.push_back(marker);
+    markers->markers.push_back(center);
 }
 
-void FrontierEvaluator::createMarkerFromVoxelStates(visualization_msgs::MarkerArray* markers) {
+void FrontierEvaluator::createMarkerFromVoxelStates(visualization_msgs::MarkerArray *markers) {
```
Comment (gajena) [inline frontier_explorer/src/frontier_evaluator.cpp]: please follow same pattern * and & with data type

File: frontier_explorer/src/frontier_evaluator.cpp [hunk 7]
```diff
@@ -60,11 +63,71 @@ FrontierEvaluator::FrontierEvaluator(ros::NodeHandle& nh, ros::NodeHandle& nh_pr
         neighbor_voxels_.push_back(Eigen::Vector3d(-vs, vs, -vs));
         neighbor_voxels_.push_back(Eigen::Vector3d(-vs, -vs, -vs));
     }
+
+    planar_neighbor_voxels_.push_back(Eigen::Vector3d(vs, 0, 0));
```
Comment (gajena) [inline frontier_explorer/src/frontier_evaluator.cpp:66]: is for will be in new PR?

File: frontier_explorer/src/frontier_evaluator.cpp [hunk 8]
```diff
@@ -118,12 +181,24 @@ bool FrontierEvaluator::getVoxelWeight(const Eigen::Vector3d& point, double& wei
...
-    size_t num_voxels_per_block = vps * vps * vps;
+void FrontierEvaluator::clusterFrontiers() {
+    while(!hash_map_.empty()) {
+        Frontier frontier;
+        neighbour_coords(hash_map_.begin()->first, frontier);
+        if(frontier.points.size() < min_frontier_size_) { continue; }
+        frontiers_.push_back(frontier);
 
+        ariitk_planning_msgs::Frontier frontier_msg;
+        frontier_msg.center.x = frontier.center.x();
```
Comment (gajena) [inline frontier_explorer/src/frontier_evaluator.cpp]: pls write a conversion funtion used more than once.

File: frontier_explorer/src/frontier_evaluator.cpp [hunk 9]
```diff
@@ -60,11 +63,71 @@ FrontierEvaluator::FrontierEvaluator(ros::NodeHandle& nh, ros::NodeHandle& nh_pr
...
+    size_t num_voxels_per_block = vps * vps * vps;
+
+    voxblox::BlockIndexList blocks;
+    esdf_server_.getTsdfMapPtr()->getTsdfLayerPtr()->getAllAllocatedBlocks(&blocks);
+    for (const auto &index : blocks) {
+        const voxblox::Block<voxblox::TsdfVoxel> &block = esdf_server_.getTsdfMapPtr()->getTsdfLayerPtr()->getBlockByIndex(index);
+        for (size_t linear_index = 0; linear_index < num_voxels_per_block; ++linear_index) {
+            Eigen::Vector3d coord = block.computeCoordinatesFromLinearIndex(linear_index).cast<double>();
+            if (isFrontierVoxel(coord)){
+                std::string hash = std::to_string(int(coord.x() / voxel_size_)) + "," + std::to_string(int(coord.y()/voxel_size_));
```
Comment (gajena) [inline frontier_explorer/src/frontier_evaluator.cpp]: make funciton

File: frontier_explorer/include/frontier_explorer/frontier_evaluator.hpp [hunk 10]
```diff
@@ -7,44 +7,74 @@
...
+
+        std::vector<Frontier> frontiers_;
+        std::vector<Eigen::Vector3d> neighbor_voxels_;
+        std::vector<Eigen::Vector3d> planar_neighbor_voxels_;
+
+        mav_planning::PhysicalConstraints constraints_;
+
         voxblox::EsdfServer esdf_server_;
 
+        void neighbour_coords(const std::string& key, Frontier& frontier) {
```
Comment (gajena) [inline frontier_explorer/include/frontier_explorer/frontier_evaluator.hpp]: move to cpp just filling space here

File: frontier_explorer/src/frontier_evaluator.cpp (hunk 7 above)
Comment (ashwin2802) [reply frontier_explorer/src/frontier_evaluator.cpp:66]: yep. should be in the branches ahead

File: frontier_explorer/src/frontier_evaluator.cpp (hunk 9 above)
Comment (ashwin2802) [reply frontier_explorer/src/frontier_evaluator.cpp]: of what exactly?

**Comment by ashwin2802** [comment]
@gajena ping

**Review by gajena — APPROVED** [review]

**Pull request merged**